In [1]:
import pandas as pd
import statistics as st


In [2]:
df=pd.read_csv("StudentsPerformance.csv")


In [3]:
df

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

In [4]:
st.mode(df['math score'])

65

# Question2

In [5]:
st.mode(df['parental level of education'])

'some college'

# Question1

In [6]:
femaledf =df[df['gender']=='female']

In [7]:
femaledf['gender'].value_counts()


female    518
Name: gender, dtype: int64

In [8]:
maledf =df[df['gender']=='male']

In [9]:
maledf['gender'].value_counts()

male    482
Name: gender, dtype: int64

# Question 3

In [10]:
st.mean(maledf['math score'])

68.82157676348548

In [11]:
maledf[maledf['math score']==68]

gender race/ethnicity parental level of education         lunch  \
734   male        group A            some high school  free/reduced   
735   male        group A                 high school      standard   
736   male        group B            some high school      standard   
737   male        group C                 high school      standard   
738   male        group C            some high school  free/reduced   
739   male        group C                some college  free/reduced   
740   male        group C          associate's degree  free/reduced   
741   male        group D           bachelor's degree      standard   
742   male        group D                some college      standard   
743   male        group D                 high school      standard   
744   male        group D           bachelor's degree  free/reduced   
745   male        group E                some college      standard   
746   male        group E           bachelor's degree      standard   
747   male        group E            some high school      standard   
748   male        group E                some college      standard   

    test preparation course  math score  reading score  writing score  
734                    none          68             72             64  
735                    none          68             70             66  
736                    none          68             54             53  
737                    none          68             60             53  
738                    none          68             63             54  
739                    none          68             68             61  
740                    none          68             65             61  
741               completed          68             74             74  
742                    none          68             59             62  
743               completed          68             64             66  
744                    none          68             68             67  
745                    none          68             60             59  
746                    none          68             68             64  
747               completed          68             51             57  
748                    none          68             72             65

In [12]:
st.mean(femaledf['math score'])

63.37644787644788

In [13]:
femaledf[femaledf['math score']==63]

gender race/ethnicity parental level of education         lunch  \
232  female        group B                some college      standard   
233  female        group B            some high school  free/reduced   
234  female        group C          associate's degree      standard   
235  female        group C           bachelor's degree      standard   
236  female        group C                some college      standard   
237  female        group C            some high school      standard   
238  female        group C                 high school      standard   
239  female        group C                some college      standard   
240  female        group C                some college  free/reduced   
241  female        group D                some college  free/reduced   
242  female        group D           bachelor's degree  free/reduced   
243  female        group D                some college      standard   
244  female        group E                some college      standard   

    test preparation course  math score  reading score  writing score  
232                    none          63             65             61  
233               completed          63             78             79  
234                    none          63             67             70  
235                    none          63             75             81  
236               completed          63             78             80  
237                    none          63             73             68  
238                    none          63             69             74  
239                    none          63             74             74  
240               completed          63             73             71  
241               completed          63             80             80  
242                    none          63             73             78  
243                    none          63             64             67  
244               completed          63             72             70

In [14]:
st.mean(maledf['reading score'])

65.54564315352697

In [15]:
maledf[maledf['reading score']==65]

gender race/ethnicity parental level of education         lunch  \
632   male        group B                some college  free/reduced   
675   male        group C          associate's degree      standard   
707   male        group B          associate's degree      standard   
716   male        group B                some college      standard   
740   male        group C          associate's degree  free/reduced   
763   male        group B                 high school      standard   
789   male        group B                 high school      standard   

    test preparation course  math score  reading score  writing score  
632               completed          59             65             66  
675                    none          62             65             58  
707               completed          65             65             63  
716                    none          66             65             60  
740                    none          68             65             61  
763                    none          70             65             60  
789               completed          72             65             68

In [16]:
st.mean(femaledf['reading score'])

72.59073359073359

In [17]:
femaledf[femaledf['reading score']==72]

gender race/ethnicity parental level of education         lunch  \
98   female        group C           bachelor's degree  free/reduced   
123  female        group C                 high school  free/reduced   
142  female        group C          associate's degree      standard   
152  female        group E             master's degree  free/reduced   
178  female        group A           bachelor's degree      standard   
187  female        group C                 high school      standard   
194  female        group D            some high school      standard   
199  female        group B                 high school  free/reduced   
200  female        group C                some college      standard   
209  female        group B           bachelor's degree      standard   
211  female        group C                 high school      standard   
222  female        group C                some college  free/reduced   
229  female        group D           bachelor's degree  free/reduced   
244  female        group E                some college      standard   
264  female        group C           bachelor's degree      standard   
287  female        group B                 high school      standard   
309  female        group D                 high school      standard   
333  female        group D                 high school      standard   
344  female        group C                some college      standard   
362  female        group B           bachelor's degree      standard   
364  female        group C                some college      standard   
383  female        group B                 high school      standard   
411  female        group D            some high school      standard   

    test preparation course  math score  reading score  writing score  
98                completed          51             72             79  
123                    none          53             72             64  
142               completed          55             72             79  
152                    none          56             72             65  
178                    none          59             72             70  
187                    none          59             72             68  
194                    none          59             72             80  
199                    none          60             72             68  
200                    none          60             72             74  
209                    none          61             72             70  
211                    none          61             72             70  
222                    none          62             72             70  
229                    none          62             72             74  
244               completed          63             72             70  
264                    none          65             72             74  
287                    none          66             72             70  
309                    none          67             72             74  
333                    none          69             72             77  
344               completed          70             72             76  
362                    none          72             72             74  
364                    none          72             72             71  
383                    none          74             72             72  
411                    none          76             72             71

In [18]:
st.mean(maledf['writing score'])

63.446058091286304

In [19]:
maledf[maledf['writing score']==63]

gender race/ethnicity parental level of education         lunch  \
692   male        group D                some college      standard   
707   male        group B          associate's degree      standard   
719   male        group D                 high school      standard   
733   male        group D          associate's degree      standard   
758   male        group D                some college  free/reduced   
796   male        group E          associate's degree      standard   
804   male        group C            some high school      standard   
887   male        group D          associate's degree      standard   

    test preparation course  math score  reading score  writing score  
692               completed          63             55             63  
707               completed          65             65             63  
719                    none          66             69             63  
733               completed          67             54             63  
758               completed          69             60             63  
796                    none          72             64             63  
804                    none          73             66             63  
887                    none          80             63             63

In [20]:
st.mean(femaledf['writing score'])

72.46718146718146

In [21]:
femaledf[femaledf['writing score']==72]

gender race/ethnicity parental level of education         lunch  \
107  female        group B            some high school  free/reduced   
147  female        group C           bachelor's degree      standard   
154  female        group B            some high school      standard   
171  female        group C                some college      standard   
186  female        group C          associate's degree      standard   
202  female        group C                 high school      standard   
212  female        group D            some high school      standard   
228  female        group D                some college      standard   
308  female        group C           bachelor's degree  free/reduced   
340  female        group B            some high school      standard   
383  female        group B                 high school      standard   
395  female        group E            some high school  free/reduced   

    test preparation course  math score  reading score  writing score  
107               completed          52             67             72  
147               completed          56             79             72  
154                    none          57             67             72  
171                    none          58             67             72  
186               completed          59             73             72  
202                    none          60             68             72  
212               completed          61             74             72  
228                    none          62             70             72  
308                    none          67             75             72  
340                    none          70             64             72  
383                    none          74             72             72  
395                    none          74             74             72

In [22]:
nonedf=df[df['test preparation course']=='none']

In [23]:
st.mean(nonedf['math score'])

63.78971962616822

In [24]:
nonedf[nonedf['math score']==63]

gender race/ethnicity parental level of education         lunch  \
232  female        group B                some college      standard   
234  female        group C          associate's degree      standard   
235  female        group C           bachelor's degree      standard   
237  female        group C            some high school      standard   
238  female        group C                 high school      standard   
239  female        group C                some college      standard   
242  female        group D           bachelor's degree  free/reduced   
243  female        group D                some college      standard   
683    male        group A          associate's degree      standard   
684    male        group A                 high school      standard   
685    male        group B           bachelor's degree      standard   
687    male        group B                 high school  free/reduced   
689    male        group C                some college  free/reduced   
691    male        group C                some college      standard   
693    male        group D                 high school  free/reduced   
694    male        group D           bachelor's degree  free/reduced   
695    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
232                    none          63             65             61  
234                    none          63             67             70  
235                    none          63             75             81  
237                    none          63             73             68  
238                    none          63             69             74  
239                    none          63             74             74  
242                    none          63             73             78  
243                    none          63             64             67  
683                    none          63             61             61  
684                    none          63             63             62  
685                    none          63             71             69  
687                    none          63             48             47  
689                    none          63             61             54  
691                    none          63             63             60  
693                    none          63             57             56  
694                    none          63             66             67  
695                    none          63             61             60

In [41]:
completeddf=df[df['test preparation course']=='completed']

In [42]:
st.mean(completeddf['math score'])

69.96648044692738

In [45]:
completeddf[completeddf['math score']== 69]

gender race/ethnicity parental level of education         lunch  \
325  female        group B                 high school      standard   
326  female        group C                some college      standard   
330  female        group C             master's degree      standard   
336  female        group D                some college      standard   
337  female        group D                 high school      standard   
338  female        group D            some high school  free/reduced   
753    male        group C                 high school      standard   
758    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
325               completed          69             76             74  
326               completed          69             90             88  
330               completed          69             84             85  
336               completed          69             79             81  
337               completed          69             77             78  
338               completed          69             86             81  
753               completed          69             58             53  
758               completed          69             60             63

In [46]:
st.mean(completeddf['reading score'])

74.17597765363128

In [47]:
completeddf[completeddf['reading score']== 74]

gender race/ethnicity parental level of education         lunch  \
157  female        group D          associate's degree  free/reduced   
212  female        group D            some high school      standard   
267  female        group C            some high school      standard   
286  female        group B           bachelor's degree      standard   
291  female        group C           bachelor's degree  free/reduced   
294  female        group E                 high school  free/reduced   
296  female        group E                some college      standard   
307  female        group C            some high school      standard   
729    male        group C                some college  free/reduced   
741    male        group D           bachelor's degree      standard   
779    male        group C           bachelor's degree      standard   
785    male        group E          associate's degree      standard   
865    male        group E          associate's degree  free/reduced   
879    male        group E           bachelor's degree  free/reduced   
903    male        group E                some college      standard   
946    male        group E                some college  free/reduced   
957    male        group D            some high school      standard   

    test preparation course  math score  reading score  writing score  
157               completed          57             74             76  
212               completed          61             74             72  
267               completed          65             74             77  
286               completed          66             74             81  
291               completed          66             74             81  
294               completed          66             74             78  
296               completed          66             74             73  
307               completed          67             74             77  
729               completed          67             74             70  
741               completed          68             74             74  
779               completed          71             74             68  
785               completed          71             74             68  
865               completed          78             74             72  
879               completed          79             74             72  
903               completed          81             74             71  
946               completed          87             74             70  
957               completed          88             74             75

In [48]:
st.mean(completeddf['writing score'])

74.68435754189944

In [49]:
completeddf[completeddf['writing score']==74]

gender race/ethnicity parental level of education         lunch  \
149  female        group D                 high school      standard   
198  female        group B            some high school      standard   
203  female        group C                 high school      standard   
251  female        group D            some high school      standard   
258  female        group A          associate's degree      standard   
325  female        group B                 high school      standard   
346  female        group D             master's degree      standard   
711    male        group D                some college      standard   
741    male        group D           bachelor's degree      standard   
768    male        group C           bachelor's degree  free/reduced   
786    male        group A                 high school      standard   
790    male        group B                some college      standard   
905    male        group B          associate's degree  free/reduced   

    test preparation course  math score  reading score  writing score  
149               completed          56             68             74  
198               completed          60             70             74  
203               completed          60             64             74  
251               completed          64             60             74  
258               completed          65             70             74  
325               completed          69             76             74  
346               completed          70             71             74  
711               completed          65             77             74  
741               completed          68             74             74  
768               completed          70             75             74  
786               completed          72             73             74  
790               completed          72             76             74  
905               completed          82             78             74

# Question4

In [25]:
maledf


gender race/ethnicity parental level of education         lunch  \
518   male        group C                 high school  free/reduced   
519   male        group A                some college  free/reduced   
520   male        group B                 high school  free/reduced   
521   male        group E            some high school      standard   
522   male        group B                some college  free/reduced   
..     ...            ...                         ...           ...   
995   male        group E                some college      standard   
996   male        group A                some college      standard   
997   male        group D                some college      standard   
998   male        group E          associate's degree  free/reduced   
999   male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
518                    none          27             34             36  
519                    none          28             23             19  
520                    none          30             24             15  
521                    none          30             26             22  
522                    none          31             32             34  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[482 rows x 8 columns]

In [26]:
st.variance(maledf['math score'])

211.88909688494752

In [27]:
import math
math.sqrt(211.89)

14.556441872930348

In [28]:
st.mean(maledf['math score'])

68.82157676348548

## 68% of males scored between 14.56 standard deviation from mean 64 for math

In [29]:
variance=st.variance(femaledf['math score'])

In [30]:

math.sqrt(variance)

16.029928046024438

In [31]:
st.mean(femaledf['math score'])

63.37644787644788

## 68% of females scored between 16.02 standard deviation from mean 63.37 for math

In [32]:
variance=st.variance(maledf['reading score'])
math.sqrt(variance)

14.149593990092226

In [33]:
st.mean(maledf['reading score'])

65.54564315352697

 68% of males scored between 14.14 standard deviation from mean 65.54 for reading

In [34]:
variance=st.variance(femaledf['reading score'])
math.sqrt(variance)

14.411017942724353

In [35]:
st.mean(femaledf['reading score'])

72.59073359073359

# 68% of females scored between 14.41 standard deviation from mean 72 for reading

In [36]:
variance=st.variance(maledf['writing score'])
math.sqrt(variance)

14.227224751791564

In [37]:
st.mean(maledf['writing score'])

63.446058091286304

# 68% of males scored between 14.22 standard deviation from mean 63 for writing

In [38]:
variance=st.variance(femaledf['writing score'])
math.sqrt(variance)

14.84484176575076

In [39]:
st.mean(femaledf['writing score'])

72.46718146718146

# 68% of females scored between 14.84 standard deviation from mean 72.46 for writing

# Top 25% of students in math score

In [50]:
q75=df['math score'].quantile(q=0.75)
q75

77.0

In [51]:
df[df['math score'].ge(q75)]

gender race/ethnicity parental level of education         lunch  \
414  female        group A                 high school  free/reduced   
415  female        group B             master's degree  free/reduced   
416  female        group B           bachelor's degree  free/reduced   
417  female        group B             master's degree      standard   
418  female        group B                 high school      standard   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
414               completed          77             88             85  
415               completed          77             97             94  
416                    none          77             85             87  
417                    none          77             90             84  
418               completed          77             82             89  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[255 rows x 8 columns]